src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java

```diff
@@ -148,4 +186,200 @@ private Student createStudent(teammates.storage.sqlentity.Course newCourse,
+            }
+        }
+
+        deleteCursorPositionFile();
+        log(isPreview() ? "Preview Completed!" : "Migration Completed!");
+        log("Total number of course entities: " + numberOfScannedKey.get());
+        log("Number of affected course entities: " + numberOfAffectedEntities.get());
+        log("Number of updated course entities: " + numberOfUpdatedEntities.get());
+    }
+
+    /**
+     * Deletes the course and its related entities from sql database.
+     */
+    private void deleteCourseCascade(Course oldCourse) {
+        // TODO: Implement deleteCourseCascase
+        log("delete course id: " + oldCourse.getUniqueId());
+    }
+
+    /**
+     * Reads the cursor position from the saved file.
+     *
+     * @return cursor if the file can be properly decoded.
+     */
+    private Optional<Cursor> readPositionOfCursorFromFile() {
+        try {
+            String cursorPosition = FileHelper.readFile(BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor");
+            return Optional.of(Cursor.fromUrlSafe(cursorPosition));
+        } catch (IOException | IllegalArgumentException e) {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Saves the cursor position to a file so it can be used in the next run.
+     */
+    private void savePositionOfCursorToFile(Cursor cursor) {
+        try {
+            FileHelper.saveFile(
+                    BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor", cursor.toUrlSafe());
+        } catch (IOException e) {
+            logError("Fail to save cursor position " + e.getMessage());
+        }
+    }
+
+    /**
+     * Deletes the cursor position file.
+     */
+    private void deleteCursorPositionFile() {
+        FileHelper.deleteFile(BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor");
+    }
+
+    /**
+     * Migrates the entity without transaction for better performance.
+     */
+    private void migrateWithoutTrx(Course entity) {
+        doMigration(entity);
+    }
+
+    /**
+     * Migrates the entity and counts the statistics.
+     */
+    private void doMigration(Course entity) {
+        try {
+            if (!isMigrationNeeded(entity)) {
+                return;
+            }
+            numberOfAffectedEntities.incrementAndGet();
+            if (!isPreview()) {
+                migrateEntity(entity);
+                numberOfUpdatedEntities.incrementAndGet();
+            }
+        } catch (Exception e) {
+            logError("Problem migrating entity " + entity);
+            logError(e.getMessage());
+        }
+    }
+
+    /**
+     * Stores the entity to save in a buffer and saves it later.
+     */
+    protected void saveEntityDeferred(BaseEntity entity) {
+        if (shouldUseTransaction()) {
+            throw new RuntimeException("Batch saving is not supported for transaction!");
+        }
+        entitiesSavingBuffer.add(entity);
```

Comment:
Will we still be flushing inside `saveEntityDeferred`?

Reply:
I think probably we need to check if the entitiesSavingBuffer has hit some size then flush?

Reply:
Yes, we discussed this and intended to use a `MAX_BUFFER_SIZE` constant, and check the buffer size in `saveEntityDeferred()`.